### PR TITLE
✨🐛♻️ Billing center

### PR DESCRIPTION
**Pull request by ignapas**
## What do these changes do?
Third iteration for the Billing center
* New autorecharge view with loading state
* Export endpoint is a GET
* Usage table
  * Column order persistence
  * Export function
  * Date filtering
  * Sorting
  * Infinite scrolling / pagination
  * Request splitting (to overcome 49 rows server limit and variable and buggy requests coming from Qlooxqlud)
* Beter general layouting (table height, column widths and more...)
* Removed border for gateway credit card view

|  before | now  |
|---|---|
| image |  image  |

image

## Related issue/s

- resolves ITISFoundation/osparc-issues#1244

## Dev Checklist

- [ ] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

**Review by matusdrobuliak66 — APPROVED** [review]
Looking good thanks! did you have a look also whether the usage listing works correctly in non-billable product? thanks

**Comment by ignapas** [comment]
> Looking good thanks! did you have a look also whether the usage listing works correctly in non-billable product? thanks

No I haven't. How do I bring my local deployment back to non billable? just change the priority for the product?

File: services/web/server/src/simcore_service_webserver/users/_preferences_models.py [hunk 1]
```diff
@@ -95,6 +95,11 @@ class TelemetryLowDiskSpaceWarningThresholdFrontendUserPreference(
     value: int = 5  # in gigabytes
 
 
+class BillingCenterUsageColumnOrderFrontendUserPreference(FrontendUserPreference):
+    preference_identifier: PreferenceIdentifier = "billingCenterUsageColumnOrder"
+    value: list[int] | None = None
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/users/_preferences_models.py:100]: Q: just to make sure, you can make this one return by default an empty list. You sure you don't need that and prefer the None.

**Review by GitHK — APPROVED** [review]
I have fixed the issue with your failing test.

**Comment by matusdrobuliak66** [comment]
> > Looking good thanks! did you have a look also whether the usage listing works correctly in non-billable product? thanks
> 
> No I haven't. How do I bring my local deployment back to non billable? just change the priority for the product?

yes, let me know if you have some issues with it

**Review by elisabettai — APPROVED** [review]
Looks great! 🤩 

I went through the requests from NK, and I think you have implemented all what he thinks are important:
1. Usage table: having a way to scroll down and see ALL the rows in the Usage Table
2. Usage table: export to csv (potentially ALL the rows).

Some small questions/suggestions:
3. Usage table: do not show the "node" column by default in the s4l products (I think most of the users will only have a sim4life node inside a project and the project name is normally more meaningful)
4. Usage table: does the  table auto-resizes if the user resizes the window?
5. Autorecharge: can the user change the threshold for triggering the autorecharge? If yes, it might make sense to add a way for changing it in the autorecharge window.

**Comment by ignapas** [comment]
> Looks great! 🤩
> 
> I went through the requests from NK, and I think you have implemented all what he thinks are important:
> 
> 1. Usage table: having a way to scroll down and see ALL the rows in the Usage Table
> 2. Usage table: export to csv (potentially ALL the rows).
> 
> Some small questions/suggestions: 3. Usage table: do not show the "node" column by default in the s4l products (I think most of the users will only have a sim4life node inside a project and the project name is normally more meaningful) 4. Usage table: does the table auto-resizes if the user resizes the window? 5. Autorecharge: can the user change the threshold for triggering the autorecharge? If yes, it might make sense to add a way for changing it in the autorecharge window.

3. I found it important to distinguish which service within the project is using credits... even if most users will have only one, those who have more than one would not know from which service it is coming...
4. No table auto-resizing for the moment... we could consider it in the next iteration as that would require me to go even deeper into the dark and smelly forests of Qlooonkloos
5. Afaik that is not something the user is allow to change @pcrespov knows better

**Comment by ignapas** [comment]
> Looks great! 🤩
> 
> I went through the requests from NK, and I think you have implemented all what he thinks are important:
> 
> 1. Usage table: having a way to scroll down and see ALL the rows in the Usage Table
> 2. Usage table: export to csv (potentially ALL the rows).
> 
> Some small questions/suggestions: 3. Usage table: do not show the "node" column by default in the s4l products (I think most of the users will only have a sim4life node inside a project and the project name is normally more meaningful) 4. Usage table: does the table auto-resizes if the user resizes the window? 5. Autorecharge: can the user change the threshold for triggering the autorecharge? If yes, it might make sense to add a way for changing it in the autorecharge window.

Although I think the user should at least be able to know what's that threshold

File: api/specs/web-server/_resource_usage.py [hunk 2]
```diff
@@ -73,8 +73,8 @@ async def list_resource_usage_services(
 )
 
 
-@router.post(
-    "/services/-/resource-usages:export",
+@router.get(
+    "/services/-/usage-report",
```
[the remark this answers is not in the document]
Comment (ignapas) [reply api/specs/web-server/_resource_usage.py:77]: imo it is still semantically sound... the good part is that the front-end does not need to know how it is solved back-end side, it is just a GET that works. and it was never handled as a long lasting operation so anyway we have to redo it if it gets too slow in the future

File: api/specs/web-server/_resource_usage.py [hunk 3]
```diff
@@ -73,8 +73,8 @@ async def list_resource_usage_services(
...
     status_code=status.HTTP_302_FOUND,
```
[the remark this answers is not in the document]
Comment (ignapas) [reply api/specs/web-server/_resource_usage.py:78]: yes, it is still a redirection atm

**Comment by elisabettai** [comment]
> > Looks great! 🤩
> > I went through the requests from NK, and I think you have implemented all what he thinks are important:
> > 
> > 1. Usage table: having a way to scroll down and see ALL the rows in the Usage Table
> > 2. Usage table: export to csv (potentially ALL the rows).
> > 
> > Some small questions/suggestions: 3. Usage table: do not show the "node" column by default in the s4l products (I think most of the users will only have a sim4life node inside a project and the project name is normally more meaningful) 4. Usage table: does the table auto-resizes if the user resizes the window? 5. Autorecharge: can the user change the threshold for triggering the autorecharge? If yes, it might make sense to add a way for changing it in the autorecharge window.
> 
> Although I think the user should at least be able to know what's that threshold

Yes, I totally agree with this. For now there's no way for the users to know when the auto-recharge will be triggered. Unless it was decided at the upper levels to be like that, I'd ask @pcrespov what is the (hard-coded?) value for the moment and display it the "Auto-recharge" window, instead of saying "when it runs low".

**Comment by matusdrobuliak66** [comment]
> > > Looks great! 🤩
> > > I went through the requests from NK, and I think you have implemented all what he thinks are important:
> > > 
> > > 1. Usage table: having a way to scroll down and see ALL the rows in the Usage Table
> > > 2. Usage table: export to csv (potentially ALL the rows).
> > > 
> > > Some small questions/suggestions: 3. Usage table: do not show the "node" column by default in the s4l products (I think most of the users will only have a sim4life node inside a project and the project name is normally more meaningful) 4. Usage table: does the table auto-resizes if the user resizes the window? 5. Autorecharge: can the user change the threshold for triggering the autorecharge? If yes, it might make sense to add a way for changing it in the autorecharge window.
> > 
> > 
> > Although I think the user should at least be able to know what's that threshold
> 
> Yes, I totally agree with this. For now there's no way for the users to know when the auto-recharge will be triggered. Unless it was decided at the upper levels to be like that, I'd ask @pcrespov what is the (hard-coded?) value for the moment and display it the "Auto-recharge" window, instead of saying "when it runs low".

its hardcoded = 100 credits

**Comment by pcrespov** [comment]
> > Yes, I totally agree with this. For now there's no way for the users to know when the auto-recharge will be triggered. Unless it was decided at the upper levels to be like that, I'd ask @pcrespov what is the (hard-coded?) value for the moment and display it the "Auto-recharge" window, instead of saying "when it runs low".
> 
> its hardcoded = 100 credits

@elisabettai @matusdrobuliak66 strictly speaking is **not** hardcoded (as of a constant value written in the code) but rather configurable via `PAYMENTS_AUTORECHARGE_MIN_BALANCE_IN_CREDITS` at an app-scope :-) i.e. it can be changed on any of the deploys right now i we want to :-)

**Comment by pcrespov** [comment]
> 2. I don't get what's the preview version thing but I WOULD NOT USE UPPERCASE! :D

@ignapas you do not have access to the mocks of appmotion? that is what I mean by the "preview version"

**Comment by ignapas** [comment]
> > 2. I don't get what's the preview version thing but I WOULD NOT USE UPPERCASE! :D
> 
> @ignapas you do not have access to the mocks of appmotion? that is what I mean by the "preview version"

Ah yes, sorry! but I don't think they made a design specific for the auto-recharge. I only more less replicated the one for the buy credits

**Comment by ignapas** [comment]
> > > Yes, I totally agree with this. For now there's no way for the users to know when the auto-recharge will be triggered. Unless it was decided at the upper levels to be like that, I'd ask @pcrespov what is the (hard-coded?) value for the moment and display it the "Auto-recharge" window, instead of saying "when it runs low".
> > 
> > 
> > its hardcoded = 100 credits
> 
> @elisabettai @matusdrobuliak66 strictly speaking is **not** hardcoded (as of a constant value written in the code) but rather configurable via `PAYMENTS_AUTORECHARGE_MIN_BALANCE_IN_CREDITS` at an app-scope :-) i.e. it can be changed on any of the deploys right now i we want to :-)
> 
> https://github.com/ITISFoundation/osparc-simcore/blob/master/services/payments/src/simcore_service_payments/core/settings.py#L81-L83

something else for the next iteration, to propagate it to the front-end... let's do a follow up after this gets merged

**Pull request merged**